packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3706,6 +3706,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   void _didChangeTextEditingValue() {
+    if (_hasFocus && !_value.selection.isValid) {
+      // If this field is focused and the selection is invalid, place the cursor at
+      // the end. Does not rely on _handleFocusChanged because it makes selection
+      // handles visible on Android.
+      // Unregister as a listener to the text controller while making the change.
+      widget.controller.removeListener(_didChangeTextEditingValue);
+      widget.controller.selection = _adjustedSelectionWhenFocused()!;
```

Comment:
I find it a little strange that _adjustedSelectionWhenFocused returns a nullable value and we use `!` here.  Like you have to have knowledge about the `if` statement above and the internals of _adjustedSelectionWhenFocused in order to use that `!`.

I can't think of a better way to do it though. If nothing jumps out at you as a way to improve that, I say don't worry about it and leave it as-is.